### PR TITLE
Rewrite pruned conditionals as their evaluated constants.

### LINKED ISSUE #3879: is None branch incorrectly pruned in numba 0.43!
The following code does not execute the is None branch in numba 0.43 on Mac OSX:

~~~python
import numpy as np
import numba

@numba.jit
def bug(a,b):
    if a.ndim == 1:
        if b is None:
            # FIXME Numba 0.43 - this branch is not executed when run using numba 0.43!
            print('should get here')
            return None
        print('************** BUG')

    return []

bug(np.arange(10),None)
~~~

Obviously, this is **VERY BAD**.

Here is my conda environment:

~~~sh
$ conda list
# packages in environment at /Users/cbarber/miniconda3/envs/numba-bug:
#
# Name                    Version                   Build  Channel
blas                      1.1                    openblas    
ca-certificates           2019.3.9             hecc5488_0    
certifi                   2019.3.9                 py36_0    
intel-openmp              2019.1                      144  
libblas                   3.8.0                     4_mkl    
libcblas                  3.8.0                     4_mkl    
libcxx                    7.0.0                h2d50403_1    
libffi                    3.2.1             h0a44026_1005    
libgfortran               3.0.1                         0    
liblapack                 3.8.0                     4_mkl    
llvm-meta                 7.0.0                         0    
llvmlite                  0.28.0           py36hf292337_0    
mkl                       2019.1                      144  
ncurses                   6.1               h0a44026_1002    
numba                     0.43.0           py36h0a44026_0    h
numpy                     1.15.4          py36hbb3c62a_1002    
openblas                  0.3.3             hdc02c5d_1001    
openssl                   1.1.1b               h1de35cc_1    
pip                       19.0.3                   py36_0    
python                    3.6.7             h8dc6b48_1004    
readline                  7.0               hcfe32e1_1001    
scipy                     1.2.1           py36_blas_openblash486cb9f_0    
setuptools                40.8.0                   py36_0    
sqlite                    3.26.0            h1765d9f_1001    
tk                        8.6.9             ha441bb4_1000    
wheel                     0.33.1                   py36_0    
xz                        5.2.4             h1de35cc_1001    
zlib                      1.2.11            h1de35cc_1004    
~~~

### PR DESCRIPTION
**Pull request by stuartarchibald**
This makes it so that when a branch is pruned, the condition
that is rewritten is written to a true/false bit that matches the
evaluated value of the condition. This ensures that an object mode
fall back compilation has IR with equivalent logic.

This bug is also caused by IR mutation being carried over into
object mode. An additional fix will be made for this problem.

Fixes #3879

### CODE REVIEW

File: numba/tests/test_analysis.py [hunk 1]
```diff
@@ -317,3 +318,60 @@ def impl(x=1000):
...
+            if a.ndim == 1:
+                if b is None:
+                    return 10
+                return 12
+            return []
+
+        self.assertEqual(bug(np.arange(10), 4), 12)
+        self.assertEqual(bug(np.arange(10), None), 10)
+        self.assertEqual(bug(np.arange(10).reshape((2, 5)), 10), [])
+        self.assertEqual(bug(np.arange(10).reshape((2, 5)), None), [])
```
Comment (sklam) [inline numba/tests/test_analysis.py]: needs to add checks that we are indeeded running in objmode.

Comment (sklam) [reply numba/tests/test_analysis.py]: perhaps something like: `self.assertFalse(bug.nopython_signatures)`

Comment (stuartarchibald) [reply numba/tests/test_analysis.py]: Thanks, good point, added in bbcb551

**Comment by sklam** [comment]
Azure passed everything already.  I don't think we need to wait for travis.

**Pull request merged**